### ISSUE TITLE
par_bridge_recursion test sometimes hanging indefinitely on ppc64le-unknown-linux-gnu

### ISSUE THREAD

**Issue by decathorpe** [rayon-rs/rayon#1142]
I have noticed that there is one test that sometimes just gets "stuck" and hangs indefinitely (an hour or longer, depending on how patient I am with cancelling the build):

```
     Running `/builddir/build/BUILD/rayon-1.9.0/target/rpm/deps/par_bridge_recursion-fe4a8a5c3a55045a`
running 1 test
test par_bridge_recursion has been running for over 60 seconds
```

For example, the total build time for rayon 1.9.0 was 4-5 minutes on x86_64, i686, s390x, but on ppc64le, this test has been stuck for almost two hours now.

**Comment by cuviper** [comment]
Maybe #1119 needs a closer look?

**Comment by decathorpe** [comment]
PS: I just remembered that this might actually be not specific to ppc64le. I have seen this test hang on other architectures too, but it might just be most likely to happen on ppc64le because that is our slowest architecture and builders are often under high load.

**Comment by arnabanimesh** [comment]
I am using a complex recursion based function. I want to run multiple instances of it across multiple cores and store the outputs in a vector of tuples. I have tested it on Windows 10 and 11 and the program is built using `Rust (version: 1.77.1)`.

The program starts by using multiple cores, but gradually the usage drops until 10% (nearly 1 core worth of CPU) is being used. The task manager shows that all the cores are being used, but not at 100% and the memory usage (2GB) remains constant throughout the run. When I run the single core version, one core is able to reach 100% and the memory usage is 100-200 MB. Single core version is 2.5x faster than the multicore version. I have tried using both `par_iter` and `par_bridge`. I also tried to play around with `max stack size per worker` but to no avail.

**Comment by U0001F3A2** [comment]
Deterministic repro — oversubscribe the pool and loop the `par_bridge_recursion` pattern (48 threads on 10 cores hangs within seconds):

```rust
use rayon::prelude::*;
use std::iter::once_with;

fn main() {
    let pool = rayon::ThreadPoolBuilder::new().num_threads(48).build().unwrap();
    loop {
        pool.broadcast(|_| {
            (0..20_000).into_par_iter()
                .flat_map(|i| once_with(move || rayon::join(move || i, move || i.to_string())).par_bridge())
                .collect::<Vec<_>>();
        });
    }
}
```

`fold_with` holds `Mutex<Fuse<Iter>>` across `iter.next()`, and here `next()` re-enters rayon via `rayon::join`. A worker that then blocks on that mutex is parked and leaves work-stealing, so the lock holder can wait inside `join` for a closure only a parked worker could run → deadlock. `sample` shows the pool blocked in `__psynch_mutexwait` inside `fold_with`. The `threads_started` guard only covers same-thread re-entry.

Fix in #1311: `try_lock` — a contending fold drops its split instead of parking, so no worker parks on the iterator mutex; the holder keeps draining and nothing is lost.

**Comment by cuviper** [comment]
Ok, using that loop I did get a deadlock and gdb backtraces.

I didn't find any threads that were recursing `par_bridge` locks, and most of them were simply blocked on a contended `par_bridge` lock or else completely idle in the thread `registry::main_loop`. The single exception in my case was this one:

```
Thread 40 (Thread 0x7f3c02e346c0 (LWP 1416758) "repro"):
#0  syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
#1  0x000055ff8d55f745 in std::sys::pal::unix::futex::futex_wait (futex=0x55ffc4e56908, expected=232, timeout=...) at library/std/src/sys/pal/unix/futex.rs:73
#2  std::sys::sync::condvar::futex::Condvar::wait_optional_timeout (self=0x55ffc4e56908, mutex=<optimized out>, timeout=...) at library/std/src/sys/sync/condvar/futex.rs:49
#3  std::sys::sync::condvar::futex::Condvar::wait (self=0x55ffc4e56908, mutex=<optimized out>) at library/std/src/sys/sync/condvar/futex.rs:33
#4  0x000055ff8d5352e5 in std::sync::poison::condvar::Condvar::wait<bool> (self=0x55ffc4e56908, guard=...) at /builddir/build/BUILD/rust-1.97.0-build/rustc-1.97.0-src/library/std/src/sync/poison/condvar.rs:128
#5  0x000055ff8d53af42 in rayon_core::sleep::Sleep::sleep<rayon_core::registry::{impl#10}::wait_until_cold::{closure_env#0}> (self=0x55ffc4e584d8, idle_state=0x7f3c02e2a130, latch=0x7f3c02e2a2b0, has_injected_jobs=...) at src/sleep/mod.rs:187
#6  0x000055ff8d53a950 in rayon_core::sleep::Sleep::no_work_found<rayon_core::registry::{impl#10}::wait_until_cold::{closure_env#0}> (self=0x55ffc4e584d8, idle_state=0x7f3c02e2a130, latch=0x7f3c02e2a2b0, has_injected_jobs=...) at src/sleep/mod.rs:106
#7  0x000055ff8d529b3a in rayon_core::registry::WorkerThread::wait_until_cold (self=0x7f3c02e33380, latch=0x7f3c02e2a2b0) at src/registry.rs:806
#8  0x000055ff8d526b8b in rayon_core::registry::WorkerThread::wait_until<rayon_core::latch::SpinLatch> (self=0x7f3c02e33380, latch=0x7f3c02e2a2a8) at src/registry.rs:775
#9  0x000055ff8d51e794 in rayon_core::join::join_context::{closure#0}<rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#0}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>> (worker_thread=0x7f3c02e33380, injected=false) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-core-1.13.0/src/join/mod.rs:157
#10 0x000055ff8d516a61 in rayon_core::registry::in_worker<rayon_core::join::join_context::{closure_env#0}<rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#0}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>>, (alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>)> (op=...) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-core-1.13.0/src/registry.rs:957
#11 0x000055ff8d51d829 in rayon_core::join::join_context<rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#0}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>> (oper_a=..., oper_b=<error reading variable: Cannot access memory at address 0xe8>) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-core-1.13.0/src/join/mod.rs:132
#12 0x000055ff8d512be3 in rayon::iter::plumbing::bridge_unindexed_producer_consumer<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer> (migrated=true, splitter=..., producer=0x7f3c02e2ede0, consumer=...) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-1.12.0/src/iter/plumbing/mod.rs:465
#13 0x000055ff8d513f46 in rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer> (context=...) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-1.12.0/src/iter/plumbing/mod.rs:467
#14 0x000055ff8d52021c in rayon_core::join::join_context::call_b::{closure#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>> (migrated=true) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-core-1.13.0/src/join/mod.rs:129
#15 0x000055ff8d506cd5 in rayon_core::job::{impl#9}::call::{closure#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon_core::join::join_context::call_b::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>>> () at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-core-1.13.0/src/job.rs:218
#16 0x000055ff8d50cbb0 in core::panic::unwind_safe::{impl#23}::call_once<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon_core::job::{impl#9}::call::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon_core::join::join_context::call_b::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>>>> (self=...) at /builddir/build/BUILD/rust-1.97.0-build/rustc-1.97.0-src/library/core/src/panic/unwind_safe.rs:275
#17 0x000055ff8d52313b in std::panicking::catch_unwind::do_call<core::panic::unwind_safe::AssertUnwindSafe<rayon_core::job::{impl#9}::call::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon_core::join::join_context::call_b::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>>>>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>> (data=0x7f3c02e2aa50) at /builddir/build/BUILD/rust-1.97.0-build/rustc-1.97.0-src/library/std/src/panicking.rs:581
#18 0x000055ff8d50f9db in __rust_try ()
#19 0x000055ff8d50e1bd in std::panicking::catch_unwind<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, core::panic::unwind_safe::AssertUnwindSafe<rayon_core::job::{impl#9}::call::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon_core::join::join_context::call_b::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>>>>> (f=...) at /builddir/build/BUILD/rust-1.97.0-build/rustc-1.97.0-src/library/std/src/panicking.rs:544
#20 std::panic::catch_unwind<core::panic::unwind_safe::AssertUnwindSafe<rayon_core::job::{impl#9}::call::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon_core::join::join_context::call_b::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>>>>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>> (f=...) at /builddir/build/BUILD/rust-1.97.0-build/rustc-1.97.0-src/library/std/src/panic.rs:359
#21 0x000055ff8d520a90 in rayon_core::unwind::halt_unwinding<rayon_core::job::{impl#9}::call::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon_core::join::join_context::call_b::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>>>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>> (func=...) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-core-1.13.0/src/unwind.rs:17
#22 0x000055ff8d506107 in rayon_core::job::JobResult<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>>::call<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon_core::join::join_context::call_b::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>>> (func=...) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-core-1.13.0/src/job.rs:218
#23 0x000055ff8d50b70e in rayon_core::job::{impl#4}::execute<rayon_core::latch::SpinLatch, rayon_core::join::join_context::call_b::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>> (this=0x7f3c02c316f0) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-core-1.13.0/src/job.rs:120
#24 0x000055ff8d536dab in rayon_core::job::JobRef::execute (self=...) at src/job.rs:64
#25 0x000055ff8d529f00 in rayon_core::registry::WorkerThread::execute (self=0x7f3c02e33380, job=...) at src/registry.rs:866
#26 0x000055ff8d529b10 in rayon_core::registry::WorkerThread::wait_until_cold (self=0x7f3c02e33380, latch=0x7f3c02e2aec0) at src/registry.rs:800
#27 0x000055ff8d526b8b in rayon_core::registry::WorkerThread::wait_until<rayon_core::latch::SpinLatch> (self=0x7f3c02e33380, latch=0x7f3c02e2aeb8) at src/registry.rs:775
#28 0x000055ff8d51edfd in rayon_core::join::join_context::{closure#0}<rayon_core::join::join::call::{closure_env#0}<i32, rayon1142::main::{closure#0}::{closure#0}::{closure#0}::{closure_env#0}>, rayon_core::join::join::call::{closure_env#0}<alloc::string::String, rayon1142::main::{closure#0}::{closure#0}::{closure#0}::{closure_env#1}>, i32, alloc::string::String> (worker_thread=0x7f3c02e33380, injected=false) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-core-1.13.0/src/join/mod.rs:157
#29 0x000055ff8d516bd4 in rayon_core::registry::in_worker<rayon_core::join::join_context::{closure_env#0}<rayon_core::join::join::call::{closure_env#0}<i32, rayon1142::main::{closure#0}::{closure#0}::{closure#0}::{closure_env#0}>, rayon_core::join::join::call::{closure_env#0}<alloc::string::String, rayon1142::main::{closure#0}::{closure#0}::{closure#0}::{closure_env#1}>, i32, alloc::string::String>, (i32, alloc::string::String)> (op=...) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-core-1.13.0/src/registry.rs:957
#30 0x000055ff8d51d863 in rayon_core::join::join_context<rayon_core::join::join::call::{closure_env#0}<i32, rayon1142::main::{closure#0}::{closure#0}::{closure#0}::{closure_env#0}>, rayon_core::join::join::call::{closure_env#0}<alloc::string::String, rayon1142::main::{closure#0}::{closure#0}::{closure#0}::{closure_env#1}>, i32, alloc::string::String> (oper_a=..., oper_b=...) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-core-1.13.0/src/join/mod.rs:132
#31 0x000055ff8d51d9f1 in rayon_core::join::join<rayon1142::main::{closure#0}::{closure#0}::{closure#0}::{closure_env#0}, rayon1142::main::{closure#0}::{closure#0}::{closure#0}::{closure_env#1}, i32, alloc::string::String> (oper_a=..., oper_b=...) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-core-1.13.0/src/join/mod.rs:105
#32 0x000055ff8d50da59 in rayon1142::main::{closure#0}::{closure#0}::{closure#0} () at src/main.rs:9
#33 0x000055ff8d5113ea in core::iter::sources::once_with::{impl#1}::next<(i32, alloc::string::String), rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}> (self=0x7f3c02e2ee00) at /builddir/build/BUILD/rust-1.97.0-build/rustc-1.97.0-src/library/core/src/iter/sources/once_with.rs:94
#34 0x000055ff8d522d80 in core::iter::adapters::fuse::{impl#11}::next<core::iter::sources::once_with::OnceWith<rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}>> (self=0x7f3c02e2ee00) at /builddir/build/BUILD/rust-1.97.0-build/rustc-1.97.0-src/library/core/src/iter/adapters/fuse.rs:381
#35 0x000055ff8d522d16 in core::iter::adapters::fuse::{impl#3}::next<core::iter::sources::once_with::OnceWith<rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}>> (self=0x7f3c02e2ee00) at /builddir/build/BUILD/rust-1.97.0-build/rustc-1.97.0-src/library/core/src/iter/adapters/fuse.rs:50
#36 0x000055ff8d5213ae in rayon::iter::par_bridge::{impl#2}::fold_with<core::iter::sources::once_with::OnceWith<rayon1142::main::{closure#0}::{closure#0}::{closure_env#0}>, rayon::iter::extend::ListVecFolder<(i32, alloc::string::String)>> (self=0x7f3c02e2ede0, folder=...) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-1.12.0/src/iter/par_bridge.rs:141
[...]
```

So part of this thread's `once_with`-`join` (31) was stolen, and it went to execute something else in the meantime (25). This got to an `iter::plumbing`-`join` (11) which was also stolen, so it has to wait. It found nothing else to steal at this point, so it's just idle waiting for completion.

Some of the other threads are blocked on the same `self=0x7f3c02e2ede0` bridge mutex, which is usually ok, but this one:

```
Thread 32 (Thread 0x7f3c01e1a6c0 (LWP 1416766) "repro"):
#0  syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
#1  0x000055ff8d5043c6 in std::sys::pal::unix::futex::futex_wait (futex=0x7f3c02e2edf8, expected=2, timeout=...) at library/std/src/sys/pal/unix/futex.rs:73
#2  std::sys::sync::mutex::futex::Mutex::lock_contended (self=0x7f3c02e2edf8) at library/std/src/sys/sync/mutex/futex.rs:61
#3  0x000055ff8d519ede in std::sys::sync::mutex::futex::Mutex::lock (self=0x7f3c02e2edf8) at /builddir/build/BUILD/rust-1.97.0-build/rustc-1.97.0-src/library/std/src/sys/sync/mutex/futex.rs:33
#4  std::sync::poison::mutex::Mutex<core::iter::adapters::fuse::Fuse<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>>>::lock<core::iter::adapters::fuse::Fuse<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>>> (self=0x7f3c02e2edf8) at /builddir/build/BUILD/rust-1.97.0-build/rustc-1.97.0-src/library/std/src/sync/poison/mutex.rs:492
#5  0x000055ff8d521237 in rayon::iter::par_bridge::{impl#2}::fold_with<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>, rayon::iter::extend::ListVecFolder<(i32, alloc::string::String)>> (self=0x7f3c02e2ede0, folder=...) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-1.12.0/src/iter/par_bridge.rs:140
#6  0x000055ff8d512aca in rayon::iter::plumbing::bridge_unindexed_producer_consumer<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer> (migrated=true, splitter=..., producer=0x7f3c02e2ede0, consumer=...) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-1.12.0/src/iter/plumbing/mod.rs:471
#7  0x000055ff8d513f46 in rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer> (context=...) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-1.12.0/src/iter/plumbing/mod.rs:467
#8  0x000055ff8d52021c in rayon_core::join::join_context::call_b::{closure#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>> (migrated=true) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-core-1.13.0/src/join/mod.rs:129
#9  0x000055ff8d506cd5 in rayon_core::job::{impl#9}::call::{closure#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon_core::join::join_context::call_b::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>>> () at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-core-1.13.0/src/job.rs:218
#10 0x000055ff8d50cbb0 in core::panic::unwind_safe::{impl#23}::call_once<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon_core::job::{impl#9}::call::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon_core::join::join_context::call_b::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>>>> (self=...) at /builddir/build/BUILD/rust-1.97.0-build/rustc-1.97.0-src/library/core/src/panic/unwind_safe.rs:275
#11 0x000055ff8d52313b in std::panicking::catch_unwind::do_call<core::panic::unwind_safe::AssertUnwindSafe<rayon_core::job::{impl#9}::call::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon_core::join::join_context::call_b::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>>>>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>> (data=0x7f3c01e142b0) at /builddir/build/BUILD/rust-1.97.0-build/rustc-1.97.0-src/library/std/src/panicking.rs:581
#12 0x000055ff8d50f9db in __rust_try ()
#13 0x000055ff8d50e1bd in std::panicking::catch_unwind<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, core::panic::unwind_safe::AssertUnwindSafe<rayon_core::job::{impl#9}::call::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon_core::join::join_context::call_b::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>>>>> (f=...) at /builddir/build/BUILD/rust-1.97.0-build/rustc-1.97.0-src/library/std/src/panicking.rs:544
#14 std::panic::catch_unwind<core::panic::unwind_safe::AssertUnwindSafe<rayon_core::job::{impl#9}::call::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon_core::join::join_context::call_b::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>>>>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>> (f=...) at /builddir/build/BUILD/rust-1.97.0-build/rustc-1.97.0-src/library/std/src/panic.rs:359
#15 0x000055ff8d520a90 in rayon_core::unwind::halt_unwinding<rayon_core::job::{impl#9}::call::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon_core::join::join_context::call_b::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>>>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>> (func=...) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-core-1.13.0/src/unwind.rs:17
#16 0x000055ff8d506107 in rayon_core::job::JobResult<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>>::call<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon_core::join::join_context::call_b::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>>> (func=...) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-core-1.13.0/src/job.rs:218
#17 0x000055ff8d50b70e in rayon_core::job::{impl#4}::execute<rayon_core::latch::SpinLatch, rayon_core::join::join_context::call_b::{closure_env#0}<alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>> (this=0x7f3c005e9230) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-core-1.13.0/src/job.rs:120
#18 0x000055ff8d536dab in rayon_core::job::JobRef::execute (self=...) at src/job.rs:64
#19 0x000055ff8d529f00 in rayon_core::registry::WorkerThread::execute (self=0x7f3c01e19380, job=...) at src/registry.rs:866
#20 0x000055ff8d529b10 in rayon_core::registry::WorkerThread::wait_until_cold (self=0x7f3c01e19380, latch=0x7f3c01e14730) at src/registry.rs:800
#21 0x000055ff8d526b8b in rayon_core::registry::WorkerThread::wait_until<rayon_core::latch::SpinLatch> (self=0x7f3c01e19380, latch=0x7f3c01e14728) at src/registry.rs:775
#22 0x000055ff8d51e794 in rayon_core::join::join_context::{closure#0}<rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#0}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>> (worker_thread=0x7f3c01e19380, injected=false) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-core-1.13.0/src/join/mod.rs:157
#23 0x000055ff8d516a61 in rayon_core::registry::in_worker<rayon_core::join::join_context::{closure_env#0}<rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#0}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>>, (alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>)> (op=...) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-core-1.13.0/src/registry.rs:957
#24 0x000055ff8d51d829 in rayon_core::join::join_context<rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#0}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>, rayon::iter::plumbing::bridge_unindexed_producer_consumer::{closure_env#1}<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>, alloc::collections::linked_list::LinkedList<alloc::vec::Vec<(i32, alloc::string::String), alloc::alloc::Global>, alloc::alloc::Global>> (oper_a=..., oper_b=<error reading variable: Cannot access memory at address 0x2>) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-core-1.13.0/src/join/mod.rs:132
#25 0x000055ff8d512be3 in rayon::iter::plumbing::bridge_unindexed_producer_consumer<&rayon::iter::par_bridge::IterParallelProducer<core::iter::sources::once_with::OnceWith<repro::main::{closure#0}::{closure#0}::{closure_env#0}>>, rayon::iter::extend::ListVecConsumer> (migrated=false, splitter=..., producer=0x7f3c01e14de0, consumer=...) at /home/jistone/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rayon-1.12.0/src/iter/plumbing/mod.rs:465
[...]
```

It was in an `iter::plumbing`-`join` that also got blocked and executed stolen work (19), the same `par_bridge` (5) that thread 40 has locked. Assuming this iter is part of the one that the other is waiting for, then neither thread can make progress. 💀 

This is consistent with what @U0001F3A2 said, but a plain `try_lock` might abandon it too easily. We could loop that a few times with exponential backoff, but it's hard to know how long is an appropriate wait time to give up. If we do set some default max time though, it could also be a tunable method on `ParallelBridge` to accomodate different workloads.

**Comment by cuviper** [comment]
For the general audience though -- please avoid rayon in your `par_bridge` input if possible!

**Comment by U0001F3A2** [comment]
Thanks for the thoughts. I did some additional digging and here's the status:

I measured how much `try_lock` actually gives up, and whether a bounded retry buys it back. On a 16-core machine (release, `fold_with` on contention):

| strategy | recursion repro | + oversubscribed | contended 5 ms fold | cheap fold (N=2M) |
| --- | --- | --- | --- | --- |
| blocking `lock()` (today) | deadlock | — | 0.45s | 0.20s |
| `try_lock` → shed | ok | ok | 2.05s | 0.02s |
| one `yield_now()`/backoff → shed | ok | ok | ~2.0s (no gain) | 0.02s |
| cooperative wait, outermost fold only | ok | livelock | 0.45s | 0.04s |

So `try_lock` giving up too easily is real, but it depends on the workload. It's about 10× faster on cheap folds (no parking overhead) and about 4.6× slower only when the fold is both expensive and contended. For the common case it comes out ahead.

The backoff idea doesn't recover that slow case, though. The lock-free window here is the downstream fold itself, so a waiter that sheds after a few tries gives up before the window opens. It recovers essentially nothing (~2.0s in the table). A large enough max-wait would help, but that's the "how long to wait" problem you mentioned, and it runs into the next point.

I did get a version that fully recovers throughput (0.45s): wait cooperatively (`yield_now()` to run the holder's stolen job, then retry), but only let the outermost `par_bridge` fold on a worker wait. Any fold re-entered through `next()`/`consume()` still sheds. That keeps it correct since nothing is dropped and a waiter never holds a bridge lock, so there's no cross-bridge cycle. However, it livelocks under oversubscription. When `yield_now()` finds nothing to run, the waiter has to spin, and with threads > cores those spinners starve the lock holders. It passes at threads ≤ cores and hangs above.

This is where the tuning can't get us out of that livelock. A busy wait can't hand its core to the holder no matter how you time it. To recover the throughput safely the waiter would have to park in the scheduler, through the `wait_until` / `CoreLatch` sleep path the backtraces are already sitting in, rather than block on the mutex or spin. That path is rayon-core-private today; `yield_now`/`yield_local` only run a single job.

So I think it comes down to two options:

- Stay with `try_lock`: correct for all inputs, faster on cheap folds, slower only on expensive-contended ones. The tunable backoff could sit on top as a best-effort knob, but it can't recover the expensive-fold case without risking that oversubscription spin.
- Add a small rayon-core primitive: a worker-registered one-shot waiter that sleeps via `wait_until` and wakes when the iterator mutex is released. It parks in the scheduler, stays stealable, keeps `try_lock`'s deadlock-freedom, and restores parallel folding. Even then it wouldn't make `next()` → `broadcast()` safe (broadcast waits for a job on every worker), so `try_lock`-style shedding would stay the fallback there.

Is the rayon-core primitive worth it here, or is `try_lock` the right place to stop for #1142? Happy to prototype it if you're open to the core change, or just finish the `try_lock` patch if not. I can share the repros/benchmarks (recursion/liveness, contended-fold, cheap-fold) if useful.
